test/checks/keyboard/banner-is-top-level.js

```diff
@@ -0,0 +1,89 @@
+
+	var checkSetup = axe.testUtils.checkSetup;
+	var shadowSupported = axe.testUtils.shadowSupport.v1;
+
+	afterEach(function () {
+		fixture.innerHTML = '';
+	});
+
+	it('should return false if banner landmark is in main element', function() {
+		var main = document.createElement('main');
+		var banner = document.createElement('div');
+		banner.setAttribute('role','banner');
+		main.appendChild(banner);
+		fixture.appendChild(main);
+		assert.isFalse(checks['banner-is-top-level'].evaluate(banner));
+	});
+
+	it('should return false if banner landmark is in main element', function () {
+		var main = document.createElement('main');
+		var banner = document.createElement('div');
+		banner.setAttribute('role','banner');
+		main.appendChild(banner);
+		fixture.appendChild(main);
+		assert.isFalse(checks['banner-is-top-level'].evaluate(banner));
+	});
+
+	it('should return false if banner landmark is in div with role main', function () {
+		var main = document.createElement('div');
+		main.setAttribute('role','main');
+		var banner = document.createElement('div');
+		banner.setAttribute('role','banner');
+		main.appendChild(banner);
+		fixture.appendChild(main);
+		assert.isFalse(checks['banner-is-top-level'].evaluate(banner));
+	});
+
+	it('should return false if header is not sectioning element and in div with role search', function () {
+		var search = document.createElement('div');
+		search.setAttribute('role','search');
+		var banner = document.createElement('header');
+		search.appendChild(banner);
+		fixture.appendChild(search);
+		assert.isFalse(checks['banner-is-top-level'].evaluate(banner));
+	});
+
+
+	it('should return true if banner landmark is not contained in another landmark', function () {
+		var banner = document.createElement('div');
+		banner.setAttribute('role','banner');
+		fixture.appendChild(banner);
+		assert.isTrue(checks['banner-is-top-level'].evaluate(banner));
+	});
+
+	it('should return true if header element is not sectioning element and not contained in landmark', function () {
+		var header = document.createElement('header');
+		fixture.appendChild(header);
+		assert.isTrue(checks['banner-is-top-level'].evaluate(header));
+	});
+
+	it('should return true if header element is in sectioning element', function () {
+		var header = document.createElement('header');
+		var article = document.createElement('div');
+		article.setAttribute('role', 'main');
+		article.appendChild(header);
+		fixture.appendChild(article);
+		assert.isTrue(checks['banner-is-top-level'].evaluate(header));
+	});
+
+	(shadowSupported ? it : xit)('should test if banner in shadow DOM is top level', function () {
+		var div = document.createElement('div');
+		var shadow = div.attachShadow({ mode: 'open' });
+		shadow.innerHTML = '<div role="banner">Banner landmark</div>';
+		var checkArgs = checkSetup(shadow.querySelector('[role=banner], header'));
+		assert.isTrue(checks['banner-is-top-level'].evaluate.apply(null, checkArgs));
```

Comment:
My hunch is that the failing case might be a better test of Shadow DOM, to ensure it's actually testing across the boundary. You could put `role="banner"` inside of Shadow DOM just like you have it, but add a violating landmark in the light DOM and check that it `isFalse`.

Reply:
Tried creating a shadow DOM test case, but the test case fails the test for a banner landmark in a shadow DOM node that is a descendant of a search landmark in the light DOM.

Can you look at our test case code?

Reply:
```
	(shadowSupported ? it : xit)('should test if banner in shadow DOM is top level', function () {
		var div = document.createElement('div');
		div.setAttribute('role','search');
		var shadow = div.attachShadow({ mode: 'open' });
		shadow.innerHTML = '<div role="banner">Banner landmark</div>';
		var checkArgs = checkSetup(shadow.querySelector('[role=banner], header'));
		assert.isFalse(checks['banner-is-top-level'].evaluate.apply(null, checkArgs));
	});
```

Reply:
I've made some changes to the testUtils for Shadow DOM, I'll push my code today so you can take a look at how that's working now.

Reply:
We tried the test again and the above test case is still is not failing.   Is there any changes we need to make in the test code above?

Reply:
You need to use the Shadow DOM-specific testUtil `shadowCheckSetup` instead of `checkSetup`. It allows you to focus on the test markup instead of the boilerplate setup like `axe._tree`, which is missing from these tests. If you rebase this branch based on `origin/develop`, you'll have the latest changes I made.

```
var shadowCheckSetup = axe.testUtils.shadowCheckSetup;

// and in your test:
var checkArgs = shadowCheckSetup(
    '<div role="search"></div>',
    '<div role="banner" id="target">Banner landmark</div>'
);
```
In the newest testUtil API, you can pass HTML strings or nodes for the light DOM and Shadow DOM content, with `id="target"` on either of them depending on what you want to test. I find passing strings to be the easiest because you don't have to do any DOM manipulation yourself.

If you were needing to assert the data returned, you'd also need to use `var checkContext = axe.testUtils.MockCheckContext()`. But the test works with the above shadow test setup method.

Reply:
Yep, that worked!